rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1030,39 +1035,65 @@ static rmw_ret_t discovery_thread_stop(rmw_dds_common::Context & common_context)
   return RMW_RET_OK;
 }
 
-static bool check_create_domain(dds_domainid_t did, rmw_localhost_only_t localhost_only_option)
+static bool check_create_domain(dds_domainid_t did, rmw_discovery_params_t * discovery_params)
 {
-  const bool localhost_only = (localhost_only_option == RMW_LOCALHOST_ONLY_ENABLED);
   std::lock_guard<std::mutex> lock(gcdds().domains_lock);
-  /* return true: n_nodes incremented, localhost_only set correctly, domain exists
+  /* return true: n_nodes incremented, discovery params set correctly, domain exists
      "      false: n_nodes unchanged, domain left intact if it already existed */
   CddsDomain & dom = gcdds().domains[did];
   if (dom.refcount != 0) {
-    /* Localhost setting must match */
-    if (localhost_only == dom.localhost_only) {
+    /* Discovery parameters must match */
+    if (rmw_discovery_params_equal(discovery_params, &dom.discovery_params)) {
       dom.refcount++;
       return true;
     } else {
       RCUTILS_LOG_ERROR_NAMED(
         "rmw_cyclonedds_cpp",
-        "rmw_create_node: attempt at creating localhost-only and non-localhost-only nodes "
-        "in the same domain");
+        "check_create_domain: attempt at creating nodes in the same domain with different "
+        "discovery parameters");
       return false;
     }
   } else {
     dom.refcount = 1;
-    dom.localhost_only = localhost_only;
-
-    /* Localhost-only: set network interface address (shortened form of config would be
-       possible, too, but I think it is clearer to spell it out completely).  Empty
-       configuration fragments are ignored, so it is safe to unconditionally append a
-       comma. */
-    std::string config =
-      localhost_only ?
-      "<CycloneDDS><Domain><General><Interfaces><NetworkInterface address=\"127.0.0.1\"/>"
-      "</Interfaces></General></Domain></CycloneDDS>,"
-      :
-      "";
+    dom.discovery_params = *discovery_params;
+
+    std::string config;
+
+    switch (discovery_params->automatic_discovery_range) {
+      case RMW_AUTOMATIC_DISCOVERY_RANGE_OFF:
+        /* Automatic discovery off: disable multicast for participant discovery. */
+        config += "<CycloneDDS><Domain><General><AllowMulticast>asm,ssm</AllowMulticast>"
```

Comment:
The [Cyclone docs](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/options.md#cycloneddsdomaingeneralallowmulticast) have this to say about default settings for `AllowMulticast`:

> "default" maps on spdp if the network is a WiFi network, on true if it is a wired network

I assume the default of only SPDP for WiFi was chosen to minimize multicast traffic over wifi without totally killing the normal discovery behavior. That would make sense since multicast over wifi has some painful issues with slowing down transmission rates and even wasting device battery.

WiFi considerations are very important in modern robotics, so I wonder if we should also test whether the device is wired or wireless and maybe just use `false` here if it's a wireless device?

For that matter, do we actually need the `asm` and `ssm` at all if they're apparently safe to turn off for WiFi devices? The docs say that including `asm` enables receiving (but not transmitting) SPDP messages via multicast. Do we need or even want that for `OFF`? Will disabling `asm` have a negative impact on the static peer discovery?

As for `ssm`, the docs don't indicate the purpose/function of `ssm`-based traffic within CycloneDDS so it's unclear to me if that's safe for us to disable or if we really need it. But again, if the default is to turn it off when using WiFi then that makes me doubt that it's actually needed.

Reply:
> I assume the default of only SPDP for WiFi was chosen to minimize multicast traffic over wifi without totally killing the normal discovery behavior.

Indeed.

> For that matter, do we actually need the asm and ssm at all if they're apparently safe to turn off for WiFi devices? The docs say that including asm enables receiving (but not transmitting) SPDP messages via multicast. Do we need or even want that for OFF? Will disabling asm have a negative impact on the static peer discovery?

If you don't want multicast (including participant discovery), then I don't see much benefit for the ROS 2 default behaviour to anything other than `false`. Specifying `asm` with fully static discovery could still make sense in combination with deliberately mapping some traffic on multicast addresses, but I don't see that as an important scenario here.

> As for ssm, the docs don't indicate the purpose/function of ssm-based traffic within CycloneDDS so it's unclear to me if that's safe for us to disable or if we really need it. But again, if the default is to turn it off when using WiFi then that makes me doubt that it's actually needed.

You won't ever notice anything about SSM unless you explicitly use multicast addresses in the SSM block. I don't see a need to worry about it here.